rex-type-system/src/unify.rs

```diff
@@ -7,14 +10,46 @@ use crate::{
     types::{ADTVariant, Type, ADT},
 };
 
-pub type Subst = HashMap<Id, Type>;
+#[derive(Clone)]
+pub struct Subst {
+    entries: HashMap<Id, Arc<Type>>,
+    changed: bool,
```

Comment:
i don't think `changed` should be a field on `Subst` because it is more error-prone to manage. for example: in the current implementation after calling unification this field can be `false` or `true` depending on the constraints that end up being unified.

but the value itself is only really relevant during unification in the first place. so the lower-level unification functions can and should just accept `did_change: &mut bool` and an input (or return it as an output, given they don't currently return anything).

then we can remove this state variable. fewer state variables is almost always a good thing, and certainly is in this case.